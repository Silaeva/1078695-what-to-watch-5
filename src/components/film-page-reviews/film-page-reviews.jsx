import React, {useEffect} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {fetchComments} from "../../store/api-actions";
import {formatDate} from "../../utils";

const FilmPageReviews = (props) => {
  const {filmId, loadComments, comments, isCommentsLoading, isCommentsLoadError} = props;

  useEffect(() => {
    loadComments(filmId);
  }, [filmId]);

  const getReviewPageContent = () => {
    if (isCommentsLoading && !isCommentsLoadError) {
      return (
        <p>Loading...</p>
      );
    } else if (isCommentsLoadError) {
      return (
        <p>Sorry, something went wrong!</p>
      );
    }

    return (
      comments.map((comment, i) => {
        return (
          <div key={`${i}-${comment.user}`} className="review">
            <blockquote className="review__quote">
              <p className="review__text">{comment.text}</p>

              <footer className="review__details">
                <cite className="review__author">{comment.user}</cite>
                <time className="review__date" dateTime="2016-12-24">{formatDate(comment.date)}</time>
              </footer>
            </blockquote>

            <div className="review__rating">{comment.rating}</div>
          </div>);
      })
    );
  };

  return (
    <div className="movie-card__reviews movie-card__row">
      <div className="movie-card__reviews-col">

        {getReviewPageContent()}

      </div>
    </div>
  );
};

FilmPageReviews.propTypes = {
  filmId: PropTypes.number.isRequired,
  comments: PropTypes.arrayOf(PropTypes.shape({
    text: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    user: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired
  })
  ),
  loadComments: PropTypes.func.isRequired,
  isCommentsLoading: PropTypes.bool.isRequired,
  isCommentsLoadError: PropTypes.bool.isRequired
};

const mapStateToProps = ({DATA}) => ({
  comments: DATA.comments,
  isCommentsLoading: DATA.isCommentsLoading,
  isCommentsLoadError: DATA.isCommentsLoadError
});

const mapDispatchToProps = (dispatch) => ({
  loadComments(id) {
    dispatch(fetchComments(id));
  }
});

export {FilmPageReviews};
export default connect(mapStateToProps, mapDispatchToProps)(FilmPageReviews);
