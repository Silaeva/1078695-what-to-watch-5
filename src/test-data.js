const films = [
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 1,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 2,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 3,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 4,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 5,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 6,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 7,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 8,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 9,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 10,
    title: `Pulp Fiction`,
    description: `The lives of two mob hitmen, a boxer, a gangster & his wife, and a pair of diner bandits intertwine in four tales of violence and redemption.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Pulp_Fiction.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/pulp-fiction.jpg`,
    genre: `Crime`,
    year: 1994,
    duration: 153,
    director: `Quentin Tarantino`,
    actors: `John Travolta, Uma Thurman, Samuel L. Jackson`,
    rating: {
      score: 1.5,
      count: 163242
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Pulp_Fiction.jpg`,
    bgColor: `#795433`,
    isFavorite: false
  },
  {
    srcVideo: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    id: 11,
    title: `Snatch`,
    description: `Unscrupulous boxing promoters, violent bookmakers, a Russian gangster, incompetent amateur robbers and supposedly Jewish jewelers fight to track down a priceless stolen diamond`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Comedy`,
    year: 2000,
    duration: 104,
    director: `Guy Ritchie`,
    actors: `Jason Statham, Brad Pitt, Benicio Del Toro`,
    rating: {
      score: 5,
      count: 54654654
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Snatch.jpg`,
    bgColor: `#FDFDFC`,
    isFavorite: true
  },
  {
    srcVideo: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
    previewVideo: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    id: 12,
    title: `War of the Worlds`,
    description: `As Earth is invaded by alien tripod fighting machines, one family fights for survival.`,
    image: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Snatch.jpg`,
    previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/snatch.jpg`,
    genre: `Adventure`,
    year: 2005,
    duration: 116,
    director: `Steven Spielberg`,
    actors: `Tom Cruise, Dakota Fanning, Tim Robbins`,
    rating: {
      score: 9,
      count: 23425
    },
    bgImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/War_of_the_Worlds.jpg`,
    bgColor: `#9B7E61`,
    isFavorite: true
  }
];

const comments = [
  {
    id: 1,
    text: `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    rating: 4,
    user: `Christina`,
    userId: 19,
    date: `2020-10-09T13:38:44.769Z`
  }, {
    id: 2,
    text: `This movie is just plain bad. There must be some big payola going round this awards season. Badly written, average acting at best, all the characters are unrelatable and inlikeable. 2 hours of my life wasted.`,
    rating: 9,
    user: `Sophie`,
    userId: 25,
    date: `2020-11-04T13:38:44.769Z`
  }, {
    id: 1,
    text: `Unfortunately we don't have a reliable way to tell the true ratings of a movie.`,
    rating: 7,
    user: `Jack`,
    userId: 11,
    date: `2020-10-09T13:38:44.769Z`
  }
];

const noop = () => {};

const genre = `Comedy`;

export {films, comments, noop, genre};
