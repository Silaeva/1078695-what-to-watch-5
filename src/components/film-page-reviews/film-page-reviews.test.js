import React from "react";
import renderer from "react-test-renderer";
// import {BrowserRouter} from "react-router-dom";
import {FilmPageReviews} from "./film-page-reviews";
import {comments, noop} from "../../test-data";

describe(`Should FilmPageReviews render correctly`, () => {
  it(`When load success`, () => {
    const tree = renderer
      .create(
        <FilmPageReviews
        filmId={1}
        loadComments={noop}
        comments={comments}
        isCommentsLoading={false}
        isCommentsLoadError={false}
        />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`When comments is loading`, () => {
    const tree = renderer
      .create(
        <FilmPageReviews
        filmId={1}
        loadComments={noop}
        comments={comments}
        isCommentsLoading={true}
        isCommentsLoadError={false}
        />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`When load error`, () => {
    const tree = renderer
      .create(
        <FilmPageReviews
        filmId={1}
        loadComments={noop}
        comments={comments}
        isCommentsLoading={false}
        isCommentsLoadError={true}
        />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
