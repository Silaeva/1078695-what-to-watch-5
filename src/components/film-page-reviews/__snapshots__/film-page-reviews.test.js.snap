// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should FilmPageReviews render correctly When comments is loading 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <p>
      Loading...
    </p>
  </div>
</div>
`;

exports[`Should FilmPageReviews render correctly When load error 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <p>
      Sorry, something went wrong! Please, try again later.
    </p>
  </div>
</div>
`;

exports[`Should FilmPageReviews render correctly When load success 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Christina
          </cite>
          <time
            className="review__date"
            dateTime="2016-12-24"
          >
            October 9, 2020
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        4
      </div>
    </div>
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          This movie is just plain bad. There must be some big payola going round this awards season. Badly written, average acting at best, all the characters are unrelatable and inlikeable. 2 hours of my life wasted.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Sophie
          </cite>
          <time
            className="review__date"
            dateTime="2016-12-24"
          >
            November 4, 2020
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        9
      </div>
    </div>
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Unfortunately we don't have a reliable way to tell the true ratings of a movie.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Jack
          </cite>
          <time
            className="review__date"
            dateTime="2016-12-24"
          >
            October 9, 2020
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        7
      </div>
    </div>
  </div>
</div>
`;
